### PR TITLE
Camera updates

### PR DESCRIPTION
**Pull request by petersteneteg**
The camera now uses ref properties instead to unify the state

### CODE REVIEW

File: include/inviwo/core/datastructures/camera.h [hunk 1]
```diff
@@ -37,7 +37,18 @@
 
 namespace inviwo {
 
-class CompositeProperty;
+namespace cameradefaults {
+constexpr vec3 lookFrom = vec3(0.0f, 0.0f, 2.0f);
+constexpr vec3 lookTo = vec3(0.0f);
+constexpr vec3 lookUp = vec3(0.0f, 1.0f, 0.0f);
+constexpr float nearPlane = 0.1f;
+constexpr float farPlane = 1000.0f;
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera.h]: This was 10000 before.

File: include/inviwo/core/datastructures/camera.h [hunk 2]
```diff
@@ -255,43 +269,41 @@ class IVW_CORE_API OrthographicCamera final : public Camera {
...
 class IVW_CORE_API SkewedPerspectiveCamera final : public Camera {
 public:
-    SkewedPerspectiveCamera(vec3 lookFrom = vec3(0.0f, 0.0f, 2.0f), vec3 lookTo = vec3(0.0f),
-                            vec3 lookUp = vec3(0.0f, 1.0f, 0.0f), float nearPlane = 0.01f,
-                            float farPlane = 10000.0f, float fieldOfView = 60.f,
-                            float aspectRatio = 1.f, vec2 frustumOffset = vec2(0.0f, 0.0f));
+    SkewedPerspectiveCamera(
+        vec3 lookFrom = cameradefaults::lookFrom, vec3 lookTo = cameradefaults::lookTo,
+        vec3 lookUp = cameradefaults::lookUp, float nearPlane = cameradefaults::nearPlane,
+        float farPlane = cameradefaults::farPlane, float fieldOfView = cameradefaults::fieldOfView,
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera.h]: Was fov = 60 intentional here since it is used for stereo views? I guess it might not matter...

File: include/inviwo/core/properties/cameraproperty.h [hunk 3]
```diff
@@ -152,21 +151,42 @@ class IVW_CORE_API CameraProperty : public CompositeProperty, public TrackballOb
 
     void invokeEvent(Event* event) override;
 
+    Property* getCameraProperty(const std::string& identifier) const;
+    void addCamerapProperty(Property* camprop);
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/cameraproperty.h]: Specify ownership

File: include/inviwo/core/datastructures/camera.h [hunk 4]
```diff
@@ -300,90 +312,99 @@ inline const vec3& Camera::getLookTo() const { return lookTo_; }
 inline const vec3& Camera::getLookUp() const { return lookUp_; }
 
 inline void Camera::setLookFrom(vec3 val) {
-    lookFrom_ = val;
-    invalidateViewMatrix();
+    if (lookFrom_ != val) {
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera.h]: Why these checks? The invalidation only sets a flag. I wonder if it makes a difference in terms of performance, do we set it to the same value many times?

File: include/inviwo/core/properties/cameraproperty.h [hunk 5]
```diff
@@ -152,21 +151,42 @@ class IVW_CORE_API CameraProperty : public CompositeProperty, public TrackballOb
...
+
+    virtual void serialize(Serializer& s) const override;
+    virtual void deserialize(Deserializer& d) override;
+
+private:
+    CameraFactory* factory_;
+    OptionPropertyString cameraType_;
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/cameraproperty.h:165]: Make public?

File: include/inviwo/core/properties/cameraproperty.h [hunk 6]
```diff
@@ -152,21 +151,42 @@ class IVW_CORE_API CameraProperty : public CompositeProperty, public TrackballOb
...
+
+    FloatVec3RefProperty lookFrom_;
+    FloatVec3RefProperty lookTo_;
+    FloatVec3RefProperty lookUp_;
+    FloatRefProperty aspectRatio_;
+    FloatRefProperty nearPlane_;
+    FloatRefProperty farPlane_;
 
 private:
+    bool changeCamera(const std::string& name);
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/cameraproperty.h:186]: Why private? Should only be done via property?

File: src/core/datastructures/camera.cpp [hunk 7]
```diff
@@ -188,68 +191,51 @@ bool PerspectiveCamera::equal(const Camera& other) const {
...
+const std::string OrthographicCamera::classIdentifier = "OrthographicCamera";
 
-    if (config == Config::Hide) {
-        if (widthProp) widthProp->setVisible(false);
-        return;
+void OrthographicCamera::updateFrom(const Camera* source) {
+    Camera::updateFrom(source);
+    if (auto oc = dynamic_cast<const OrthographicCamera*>(source)) {
+        setWidth(oc->getWidth());
+    } else if (auto pc = dynamic_cast<const PerspectiveCamera*>(source)) {
```
Comment (ResearchDaniel) [inline src/core/datastructures/camera.cpp]: What about skewedperspectiveCamera?

File: src/core/datastructures/camera.cpp [hunk 8]
```diff
@@ -99,87 +103,86 @@ vec3 Camera::getNormalizedDeviceFromNormalizedScreenAtFocusPointDepth(
...
-        return false;
+const std::string PerspectiveCamera::classIdentifier = "PerspectiveCamera";
+
+void PerspectiveCamera::updateFrom(const Camera* source) {
+    Camera::updateFrom(source);
+    if (auto pc = dynamic_cast<const PerspectiveCamera*>(source)) {
+        setFovy(pc->getFovy());
+    } else if (auto oc = dynamic_cast<const OrthographicCamera*>(source)) {
+        setFovy(glm::degrees(
+            2.0f * std::atan(oc->getWidth() / 2.0f / glm::distance(getLookTo(), getLookFrom()))));
```
Comment (ResearchDaniel) [inline src/core/datastructures/camera.cpp]: What about skewedperspectiveCamera?

File: src/core/properties/cameraproperty.cpp [hunk 9]
```diff
@@ -53,19 +54,45 @@ CameraProperty::CameraProperty(const std::string& identifier, const std::string&
...
-              InvalidationLevel::InvalidOutput, PropertySemantics::SpinBox)
-    , aspectRatio_("aspectRatio", "Aspect Ratio", 1.0f, 0.0f, std::numeric_limits<float>::max(),
-                   0.01f, InvalidationLevel::InvalidOutput, PropertySemantics::Text)
-    , nearPlane_("near", "Near Plane", 0.1f, 0.001f, 10.f, 0.001f)
-    , farPlane_("far", "Far Plane", 100.0f, 1.0f, 1000.0f, 1.0f)
+    , lookFrom_(
+          "lookFrom", "Look from", [this]() { return camera_->getLookFrom(); },
+          [this](const vec3& val) { camera_->setLookFrom(val); },
+          {-vec3(100.0f), ConstraintBehavior::Ignore}, {vec3(100.0f), ConstraintBehavior::Ignore},
+          vec3(0.1f), InvalidationLevel::InvalidOutput, PropertySemantics{"SphericalSpinBox"})
```
Comment (ResearchDaniel) [inline src/core/properties/cameraproperty.cpp]: I hate the spherical spinbox :)
The rotations in that space does not make much sense for camera interaction

File: src/core/properties/cameraproperty.cpp [hunk 10]
```diff
@@ -53,19 +54,45 @@ CameraProperty::CameraProperty(const std::string& identifier, const std::string&
...
+          {std::numeric_limits<float>::max(), ConstraintBehavior::Immutable}, 0.01f,
+          InvalidationLevel::InvalidOutput, PropertySemantics::Text)
+    , nearPlane_(
+          "near", "Near Plane", [this]() { return camera_->getNearPlaneDist(); },
+          [this](const float& val) { camera_->setNearPlaneDist(val); },
+          {0.001f, ConstraintBehavior::Ignore}, {10.0f, ConstraintBehavior::Ignore}, 0.001f)
+    , farPlane_(
+          "far", "Far Plane", [this]() { return camera_->getFarPlaneDist(); },
+          [this](const float& val) { camera_->setFarPlaneDist(val); },
+          {1.0f, ConstraintBehavior::Ignore}, {1000.0f, ConstraintBehavior::Ignore}, 1.0f)
```
Comment (ResearchDaniel) [inline src/core/properties/cameraproperty.cpp]: Increase max bound. Default of it regular value should be 10000...

File: src/core/properties/cameraproperty.cpp [hunk 11]
```diff
@@ -244,20 +217,55 @@ CameraProperty::operator const Camera&() const { return *camera_; }
...
+
+CameraProperty& CameraProperty::setCamera(const std::string& cameraIdentifier) {
+    cameraType_.setSelectedIdentifier(cameraIdentifier);
+    return *this;
+}
+
+CameraProperty& CameraProperty::setCamera(std::unique_ptr<Camera> newCamera) {
+    if (newCamera) {
+        NetworkLock lock(this);
+        hideConfiguredProperties();
```
Comment (ResearchDaniel) [inline src/core/properties/cameraproperty.cpp:227]: Should the previous visibility state be restored in the end here?

File: src/core/properties/cameraproperty.cpp [hunk 12]
```diff
@@ -355,6 +384,37 @@ void CameraProperty::invokeEvent(Event* event) {
...
+}
+
+void CameraProperty::serialize(Serializer& s) const {
+    CompositeProperty::serialize(s);
+    s.serialize("Camera", camera_);
+
+}
+
+void CameraProperty::deserialize(Deserializer& d) {
+    hideConfiguredProperties();
```
Comment (ResearchDaniel) [inline src/core/properties/cameraproperty.cpp]: Why this hiding? Might want to clarify it’s purpose.

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Wow, a small step for the camera but a giant leap for Inviwo :)

File: include/inviwo/core/datastructures/camera.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: It was 0.01 to 10000 before that range might be to big for a float and might give precision issues, that’s why I decreased the default. The automatic fitting also makes this easier to adjust

File: include/inviwo/core/datastructures/camera.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: Think 60 was just a bad choice to begin with. 38 is recommended by Martin :)

Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: It was 60 also for the perspective one, so it should not have anything to do with the stereo aspects

File: include/inviwo/core/properties/cameraproperty.h (hunk 3 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/cameraproperty.h]: Will do

File: include/inviwo/core/datastructures/camera.h (hunk 4 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: Mostly to be consistent. We set a flag but then also have to recalculate the matrices...

File: include/inviwo/core/properties/cameraproperty.h (hunk 5 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/cameraproperty.h:165]: Maybe possible now... there are a few invariants that has to be taken care of...

File: include/inviwo/core/properties/cameraproperty.h (hunk 6 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/cameraproperty.h:186]: There is a different one that is public

File: src/core/datastructures/camera.cpp (hunk 7 above)
Comment (petersteneteg) [reply src/core/datastructures/camera.cpp]: Yes could add that to

File: src/core/datastructures/camera.cpp (hunk 8 above)
Comment (petersteneteg) [reply src/core/datastructures/camera.cpp]: Yes could add that

File: src/core/properties/cameraproperty.cpp (hunk 9 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp]: xyz is not much better...

File: src/core/properties/cameraproperty.cpp (hunk 10 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp]: Why this is only for the slider...

File: src/core/properties/cameraproperty.cpp (hunk 11 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp:227]: Configure will show them again

File: src/core/properties/cameraproperty.cpp (hunk 12 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp]: Mmm will add some comment

File: include/inviwo/core/datastructures/camera.h (hunk 2 above)
Comment (ResearchDaniel) [reply include/inviwo/core/datastructures/camera.h]: http://gameaccessibilityguidelines.com/if-the-game-uses-field-of-view-3d-engine-only-set-an-appropriate-default-for-the-expected-viewing-environment/

60 is default in Unity as well

Comment (ResearchDaniel) [reply include/inviwo/core/datastructures/camera.h]: Just some other input on the question
https://forum.unity.com/threads/what-is-a-normal-field-of-view.70202/
” For PC games, the screen aspect is important, because players sit roughly at the same distance from the monitor. Let's say now you're using a 21:9 monitor, compared to a 16:9 one, you should see more space horizontally, NOT vertically, right? But in Unity, the aspect remains the same! So if you increase the FOV, you see more space horizontally AND vertically, that's the reason why developers feel weird and uncomfortable when they set the value to 90 or even higher on their camera. In fact, a higher FOV in Unity will make you feel like you're traveling at the speed of light, not only "wider view".

Last question, how to see wider? Change the aspect instead.”

Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: I think, there is a big difference between a 3D game (like a FPS) and a scientific visualization. I find that 60° create ways too much distortion. @martinfalk might also have opinions here. Also you can always change the camera to fit your preference...

File: include/inviwo/core/datastructures/camera/cameratools.h [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <memory>
+
+namespace inviwo {
+
+class CameraProperty;
+
+namespace util {
+
+IVW_CORE_API FloatRefProperty* getCameraFovProperty(CameraProperty* comp);
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/cameratools.h]: Should state that null is returned if it does not exist. 
Should be const Camera*

File: include/inviwo/core/datastructures/camera/cameratools.h [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+IVW_CORE_API std::unique_ptr<FloatRefProperty> createCameraFovProperty(
+    std::function<float()> get, std::function<void(const float&)> set);
+
+IVW_CORE_API FloatRefProperty* updateOrCreateCameraFovProperty(
+    CameraProperty* comp, std::function<float()> get, std::function<void(const float&)> set);
+
+
+IVW_CORE_API FloatRefProperty* getCameraWidthProperty(CameraProperty* comp);
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/cameratools.h]: Same as comment above. Should also state that this property is only valid for orthographic camera?

File: include/inviwo/core/datastructures/camera/cameratools.h [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+IVW_CORE_API std::unique_ptr<FloatRefProperty> createCameraWidthProperty(
+    std::function<float()> get, std::function<void(const float&)> set);
+
+IVW_CORE_API FloatRefProperty* updateOrCreateCameraWidthProperty(
+    CameraProperty* comp, std::function<float()> get, std::function<void(const float&)> set);
+
+
+
+IVW_CORE_API FloatVec2RefProperty* getCameraSeperationProperty(CameraProperty* comp);
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/cameratools.h]: I did not see (eye)separation in any cameras. Is this property only related the the CameraProperty? I think that it would be good to document where each of these properties stem from, i.e. which Camera types

File: include/inviwo/core/datastructures/camera/skewedperspectivecamera.h [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+namespace inviwo {
+
+/**
+ * \brief Camera with off axis perspective projection.
+ *
+ * The camera with unsymmetrical frustum for stereo in VR
+ * Kooima, Robert. "Generalized perspective projection." School of Elect. Eng. and Computer Science
+ * (2008): 1-7.
+ * @see Camera
+ * @see SkewedPerspectiveCamera
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/skewedperspectivecamera.h]: This Reference to itself should be removed

File: include/inviwo/core/datastructures/camera/cameratools.h (hunk 15 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera/cameratools.h]: It is also called offset in the skewedprespective camera

File: include/inviwo/core/datastructures/camera/cameratools.h (hunk 13 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera/cameratools.h]: it tries to find it in the CameraProperty, not in the camera. Guess it could be const syntactically, but semantically I don't think it should be since it returns a non-const property

File: include/inviwo/core/datastructures/camera/cameratools.h (hunk 14 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera/cameratools.h]: It works for any CameraProperty. but it would only be needed by an orthographic camera.

File: include/inviwo/core/datastructures/camera/cameratools.h (hunk 15 above)
Comment (ResearchDaniel) [reply include/inviwo/core/datastructures/camera/cameratools.h]: Aha, eye separation is in my opinion more descriptive. Would be good if our naming is consistent in the different places at least.

File: include/inviwo/core/datastructures/camera/cameratools.h (hunk 14 above)
Comment (ResearchDaniel) [reply include/inviwo/core/datastructures/camera/cameratools.h]: Ok, but later you have createOrUpdate functions indicating that it might return null (i.e. not work)

File: src/core/properties/cameraproperty.cpp (hunk 9 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp]: maybe it would be better to use xyz and then add some more buttons (?) for various view changes? (zoom+-, various rotations)

Comment (ResearchDaniel) [reply src/core/properties/cameraproperty.cpp]: Definitely! Rotate left/right/up/down, translate left/right (perhaps forward/backward) and zoom in/out

File: src/core/datastructures/camera/skewedperspectivecamera.cpp [hunk 17]
```diff
@@ -113,11 +160,11 @@ mat4 SkewedPerspectiveCamera::calculateProjectionMatrix() const {
 
 void SkewedPerspectiveCamera::serialize(Serializer& s) const {
     Camera::serialize(s);
-    s.serialize("fovy", fovy_);
+    s.serialize("fov", fovy_);
```
Comment (ResearchDaniel) [inline src/core/datastructures/camera/skewedperspectivecamera.cpp:167]: No. We have two field of views, one vertical and one horizontal (derived from the aspect ratio if I recall correctly). This is the vertical field of view. Could be better named

Comment (petersteneteg) [reply src/core/datastructures/camera/skewedperspectivecamera.cpp:167]: only one is ever stored. the other one is never used, and if it would be, it would be computed with the vertical fov and aspect ratio. Renaming this makes it much easier to convert old workspaces since the vertical fov property has always been serialised as "fov" the camera it self has never been serialised before, so changing this here will not break anything.

**Comment by petersteneteg** [comment]
How would I add my SGCT camera then? should that go in the core?

File: src/core/datastructures/camera/skewedperspectivecamera.cpp (hunk 17 above)
Comment (ResearchDaniel) [reply src/core/datastructures/camera/skewedperspectivecamera.cpp:167]: ok

**Comment by ResearchDaniel** [comment]
> How would I add my SGCT camera then? should that go in the core?

I was not aware that we had more cameras. Probably not, I guess it depends on how specialized it is to SGCT.

**Comment by petersteneteg** [comment]
that's the reason I went down this rabbit hole ;)

File: src/core/properties/cameraproperty.cpp (hunk 9 above)
Comment (martinfalk) [reply src/core/properties/cameraproperty.cpp]: careful here! Zoom means you change the fov but not the position. You just ensure that an object at a certain distance, i.e. the look at, stays about the same size on the screen.
Moving the camera closer/further away is a dolly.

File: include/inviwo/core/datastructures/camera.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: sketchup uses a default of 35°

Comment (martinfalk) [reply include/inviwo/core/datastructures/camera.h]: A vertical fov of 38° is closer to human vision. Notice the extreme perspective distortions happening at 60 degrees. And I am with Peter here that visualization is _not_ the same as games. In games you want to see as much of your surroundings as possible. Here we mostly have a tiny (square) canvas showing one dataset. Another thing to consider: to be more accurate one should actually use orthographic. But that messes with 3D perception of shapes/objects.

File: include/inviwo/core/datastructures/camera/camera.h [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+namespace inviwo {
+
+namespace cameradefaults {
+constexpr vec3 lookFrom = vec3(0.0f, 0.0f, 2.0f);
+constexpr vec3 lookTo = vec3(0.0f);
+constexpr vec3 lookUp = vec3(0.0f, 1.0f, 0.0f);
+constexpr float nearPlane = 0.1f;
+constexpr float farPlane = 1000.0f;
+constexpr float fieldOfView = 38.f;
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/camera.h:45]: Ok, but I think that we should state why we choose this

File: include/inviwo/core/datastructures/camera.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera.h]: and orthographic "zoom" (changing width) now also works with the mouse

File: src/core/properties/cameraproperty.cpp (hunk 9 above)
Comment (petersteneteg) [reply src/core/properties/cameraproperty.cpp]: on the other hand we really don't want to zoom using FOV since that will also be quite confusing in many cases...

File: include/inviwo/core/datastructures/camera/camera.h [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+    const vec3& getLookTo() const;
+    virtual void setLookTo(vec3 val);
+
+    const vec3& getLookUp() const;
+    virtual void setLookUp(vec3 val);
+
+    virtual float getAspectRatio() const;
+    virtual void setAspectRatio(float val);
+
+    virtual void zoom(float factor, std::optional<mat4> boundingBox) = 0;
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/camera/camera.h:96]: Should this functionality be in the camera class? It breaks the single responsibility principle in my opinion. Is it possible to move this the camerautil?

**Review by ResearchDaniel — APPROVED** [review]
Some unresolved comments remaining that are “nice to have”

File: include/inviwo/core/datastructures/camera/camera.h (hunk 19 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera/camera.h:96]: problem in that different cameras "zoom" differently, our perspective camera does a "dolly" moving closer to the look to point, where as a orthtographic camera changes it viewport width. Hence it is not so easy having it in a util class...

File: include/inviwo/core/datastructures/camera/camera.h (hunk 18 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/camera/camera.h:45]: as of now this selected to break as few regressions as possible. Might add do a different PR with update to just these values...
